File: apps/api/core/timestamps.py
from math import floor
from time import time


def now_unix_seconds() -> int:
    return floor(time())


def to_db_unix_seconds(ts: int) -> int:
    return ts


def from_db_unix_seconds(ts: int) -> int:
    return ts


def nullable_from_db_unix_seconds(ts: int | None) -> int | None:
    return ts


def iso_to_unix_seconds(iso: str) -> int:
    from datetime import datetime

    dt = datetime.fromisoformat(iso.replace("Z", "+00:00"))
    return floor(dt.timestamp())

File: apps/api/src/platform/timestamps.ts
/**
 * App-owned timestamps are Unix **seconds** everywhere — in the database, on
 * the wire, in the SDK contracts. Never milliseconds, never a Date, never an
 * ISO string. See CLAUDE.md → API Contracts.
 */
export function nowUnixSeconds(): number {
  return Math.floor(Date.now() / 1000)
}

export function isoToUnixSeconds(iso: string): number {
  const ms = Date.parse(iso)
  if (Number.isNaN(ms)) throw new Error(`Invalid ISO timestamp: ${iso}`)

  return Math.floor(ms / 1000)
}

/**
 * Prisma maps Postgres `BIGINT` to `bigint`. Every timestamp column in this
 * database is a Unix second count that fits in a double many times over, so the
 * conversion is lossless — but it has to happen at the serializer boundary or
 * `JSON.stringify` throws "Do not know how to serialize a BigInt".
 */
export function fromDbUnixSeconds(value: bigint): number {
  return Number(value)
}

export function nullableFromDbUnixSeconds(value: bigint | null): number | null {
  return value === null ? null : Number(value)
}

export function toDbUnixSeconds(value: number): bigint {
  return BigInt(value)
}

export function nullableToDbUnixSeconds(
  value: number | null | undefined
): bigint | null {
  return value === null || value === undefined ? null : BigInt(value)
}
